ocaml/xapi/extauth_plugin_ADwinbind.ml

```diff
@@ -31,6 +45,255 @@ let tdb_tool = !Xapi_globs.tdb_tool
 
 let debug_level = !Xapi_globs.winbind_debug_level |> string_of_int
 
+let ntlm_auth uname passwd : (unit, exn) result =
+  try
+    let args = ["--username"; uname] in
+    let _stdout =
+      Helpers.call_script ~log_output:Never ~stdin:passwd
+        !Xapi_globs.ntlm_auth_cmd args
+    in
+    Ok ()
+  with e -> Error e
+
+let call_ldap (query : string) : (string, exn) result =
+  try
+    let args = ["ads"; "search"; "--machine-pass"; query] in
+    let stdout =
+      Helpers.call_script ~log_output:On_failure !Xapi_globs.net_cmd args
+    in
+    Ok stdout
+  with e -> Error e
+
+let call_wbinfo (args : string list) : (string, exn) result =
+  try
+    (* we trust wbinfo will not print any sensitive info on failure *)
+    let stdout = Helpers.call_script ~log_output:On_failure wb_cmd args in
+    Ok stdout
+  with e -> Error e
+
+module Ldap = struct
+  type user = {
+      upn: string
+    ; account_disabled: bool
+    ; account_expired: bool
+    ; account_locked: bool
+    ; password_expired: bool
+  }
+  [@@deriving rpcty]
+
+  let string_of_user x =
+    Rpcmarshal.marshal user.Rpc.Types.ty x |> Jsonrpc.to_string
+
+  let parse_user stdout : (user, string) result =
+    (* there are two steps here:
+     * 1. parse stdout to a (string, string) map
+     * 2. calculate user details using the map *)
+    let module Map = Map.Make (String) in
+    let module P = struct
+      open Angstrom
+
+      let space = char ' '
+
+      let not_spaces = take_while @@ function ' ' -> false | _ -> true
+
+      let is_whitespace = function
+        | ' ' | '\n' | '\t' | '\r' ->
+            true
+        | _ ->
+            false
+
+      let ws = skip_while is_whitespace
+
+      let header =
```

Comment:
Maybe that is because I am not familiar with `Angstrom`,  The command result parse looks quite complicated to me.

why not just reuse the `xapi_cmd_result` and have `:` as the `sep`, add lines to the resultset, 
if the resultset is empy, then no replies, otherwise, just look up the key in the resultset?

Reply:
My issue with xapi_cmd_result is that is lenient if it can't parse a line - it simply ignores the error and moves on. Additionally I specifically want to make sure that all of the key value pairs are in one 'block'. Angstrom is also convenient for parsing the header

I agree it looks quite complicated, which is a shame

Reply:
we can ignore the headers,  and just parse the values, 
If you do care about the record numbers, perhaps just check whether the resultset is empty. 
If you does not found the target key/value pair, then you can report error.

Just a little bit worried needs to get familiar with `Angstrom` to maintain the codes.  you can make the decision.